### PR TITLE
Automatically decode x-ms-client-principal header

### PR DESCRIPTION
**Pull request by hossam-nasr**
Automatically decode `x-ms-client-principal` header #421:

- Add `HttpRequestUser` interface and `HttpRequestUserType` type, according to the definitions in #421 
- Add `user` property in `HttpRequest` interface
- Based on data in` x-ms-client-principal` header, figure out the type of authentication used, and add appropriate user field in the `HttpRequest` constructor

### CODE REVIEW

File: src/http/Request.ts [hunk 1]
```diff
@@ -19,6 +20,29 @@ export class Request extends RequestProperties {
     constructor(httpInput: RequestProperties) {
         super();
         Object.assign(this, httpInput);
+        if (httpInput.headers['x-ms-client-principal']) {
```
Comment (ejizba) [inline src/http/Request.ts]: If there's any issues parsing the client principal, I don't think that should throw an error here (basically preventing the entire execution from happening). Instead, how about we lazy-parse this by putting it in a getter - that way only people that try to access the user property would see the error.

Only downside to a getter is by default it would re-do the parsing each time. However, we can get around that pretty easily by just caching the result of our parse. Something like this:
```typescript
private _cachedUser?: HttpRequestUser;
public get user(): HttpRequestUser | null {
    if (this._cachedUser === undefined) {
        this._cachedUser = // do all your stuff to parse user
    }

    return this._cachedUser;
}
```

File: src/http/Request.ts [hunk 2]
```diff
@@ -19,6 +20,29 @@ export class Request extends RequestProperties {
     constructor(httpInput: RequestProperties) {
         super();
         Object.assign(this, httpInput);
+        if (httpInput.headers['x-ms-client-principal']) {
+            const claimsPrincipalData = JSON.parse(
+                Buffer.from(httpInput.headers['x-ms-client-principal'], 'base64').toString('utf-8')
+            );
+
+            if (claimsPrincipalData['identityProvider']) {
+                this.user = {
```
Comment (ejizba) [inline src/http/Request.ts]: Please add a few unit tests. Add at least one real-life test for both "StaticWebApps" and "AppService" (making sure to obfuscate any important secrets/keys) and then add any edge-case tests as you see fit.

File: types/index.d.ts [hunk 3]
```diff
@@ -131,6 +160,11 @@ declare module '@azure/functions' {
          * Route parameter keys and values.
          */
         params: HttpRequestParams;
+        /**
+         *  Object representing logged-in user, either through
+         *  AppService/Functions authentication, or SWA Authentication
+         */
+        user?: HttpRequestUser | null;
```
Comment (ejizba) [inline types/index.d.ts]: We have to be careful about `undefined` vs `null`, especially because we do not have linting properly set up to cover this area yet. The differences are subtle, but definitely there. In this case, I would get rid of the question mark.
```suggestion
        user: HttpRequestUser | null;
```

`undefined` means we have not figured out what the value of user is yet, whereas `null` means we have figured it out and there is no user. We should never return `undefined` here and the types should reflect that

File: src/http/ExtractHttpUser.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License.
+
+import { HttpRequestHeaders, HttpRequestUser } from '@azure/functions';
+
+export function extractHttpUserFromHeaders(headers: HttpRequestHeaders): HttpRequestUser | null {
```
Comment (ejizba) [inline src/http/ExtractHttpUser.ts]: As a general rule (which is not being followed in this repo yet), the file name should match the main exported function name. Aka rename it to `extractHttpUserFromHeaders.ts`

File: test/ExtractHttpUserTests.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        };
+
+        const headers: HttpRequestHeaders = {
+            'x-ms-client-principal': Buffer.from(JSON.stringify(claimsPrinicipalData)).toString('base64'),
+        };
+
+        const user: HttpRequestUser | null = extractHttpUserFromHeaders(headers);
+
+        expect(user).to.not.be.null;
+        expect(user?.id).to.equal(id);
```
Comment (ejizba) [inline test/ExtractHttpUserTests.ts:71]: You should also verify `user.type`.

In general, validate at least every property you added in the `types/index.d.ts` file

File: test/ExtractHttpUserTests.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            'x-ms-client-principal': Buffer.from(JSON.stringify(claimsPrincipalData)).toString('base64'),
+        };
+
+        const user: HttpRequestUser | null = extractHttpUserFromHeaders(headers);
+
+        expect(user).to.not.be.null;
+        expect(user?.id).to.equal(id);
+        expect(user?.username).to.equal(username);
+        expect(user?.identityProvider).to.equal(provider);
+        expect(user?.claimsPrincipalData).to.eql(claimsPrincipalData);
```
Comment (ejizba) [inline test/ExtractHttpUserTests.ts]: nit: Not a big fan of `eql`. I think `.to.deep.equal` does the same thing and is more obvious

File: test/ExtractHttpUserTests.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+describe('Extract Http User Claims Principal from Headers', () => {
+    it('Correctly parses AppService headers', () => {
+        const username = 'test@example.com';
+        const id: string = uuid();
+        const provider = 'aad';
+        const claimsPrincipalData = {
+            auth_typ: provider,
+            claims: [
+                { typ: 'aud', val: uuid() },
+                { typ: 'iss', val: `https://login.microsoftonline.com/${uuid()}/v2.0` },
```
Comment (ejizba) [inline test/ExtractHttpUserTests.ts]: For the sake of readability, you can delete some of this junk in the "claims" array. I know I said make the unit test as real as possible, but often the real data is way too verbose and we can get away with trimming it down a bit. Just keep the general structure as-is and obviously keep any important fields

File: types/index.d.ts [hunk 8]
```diff
@@ -131,6 +160,11 @@ declare module '@azure/functions' {
          * Route parameter keys and values.
          */
         params: HttpRequestParams;
+        /**
+         *  Object representing logged-in user, either through
```
Comment (ejizba) [inline types/index.d.ts:164]: You should ideally describe when `null` is returned. Something like https://github.com/Azure/azure-functions-nodejs-worker/pull/486/files#diff-093ad82a25aee498b11febf1cdcb6546e4d223ffcb49ed69cc275ac27ce0ccceR152

**Review by ejizba — COMMENTED** [review]
Wrote most of these comments a while ago, not sure why I left them in a "pending" state. Also please update with the latest v3.x

File: src/http/ExtractHttpUser.ts (hunk 4 above)
Comment (hossam-nasr) [reply src/http/ExtractHttpUser.ts]: Renamed the file

File: test/ExtractHttpUserTests.ts (hunk 5 above)
Comment (hossam-nasr) [reply test/ExtractHttpUserTests.ts:71]: Added checking `user.type` for both tests

File: test/ExtractHttpUserTests.ts (hunk 6 above)
Comment (hossam-nasr) [reply test/ExtractHttpUserTests.ts]: Done

File: test/ExtractHttpUserTests.ts (hunk 7 above)
Comment (hossam-nasr) [reply test/ExtractHttpUserTests.ts]: Thanks! Done

File: types/index.d.ts (hunk 8 above)
Comment (hossam-nasr) [reply types/index.d.ts:164]: Done

**Review by ejizba — APPROVED** [review]
